common/workunit/workunit.cpp

```diff
@@ -2657,85 +2660,187 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    filter.addScopeType(SSTglobal).addScopeType(SSTworkflow).addScopeType(SSTgraph);
+    const unsigned numStats = aggregateKindsMapping.numStatistics();
+    for (unsigned i=0; i<numStats; ++i)
+        filter.addOutputStatistic(aggregateKindsMapping.getKind(i));
+    filter.setDepth(1,3); // 1=global, 2=workflow, 3=graph
+    filter.setSources(SSFsearchGlobalStats);
+    filter.setIncludeNesting(0);
+    filter.finishedFilter();
+
+    StatsCollectionAggregatesLoader aggregatesLoader(statsCollection);
+    Owned<IConstWUScopeIterator> iter = &workunit.getScopeIterator(filter);
+    ForEach(*iter)
+        iter->playProperties(aggregatesLoader);
+}
+
+// getCollectionForUpdate() returns IStatisticCollection for the given subgraph
+// if clearStats==true then the existing stats are cleared for the given scope and descendants
+IStatisticCollection * GlobalStatisticCollection::getCollectionForUpdate(StatisticCreatorType creatorType, const char * creator, unsigned wfid, const char *graphName, unsigned sgId, bool clearStats)
+{
+    bool wasCreated;
+
+    StatsScopeId graphScopeId;
+    verifyex(graphScopeId.setScopeText(graphName));
+    StatsScopeId wfScopeId(SSTworkflow, wfid);
+    StatsScopeId sgScopeId(SSTsubgraph, sgId);
+
+    IStatisticCollection * sgScopeCollection = statsCollection->ensureSubScopePath({wfScopeId,graphScopeId, sgScopeId}, wasCreated);
+    if (clearStats && !wasCreated)
+        sgScopeCollection->clearStats();
+    sgScopeCollection->markDirty();
+    return createRootStatisticCollection(creatorType, creator, wfScopeId, graphScopeId, sgScopeCollection);
+}
+
+// Re-calculated aggregates from subgraph scope
+bool GlobalStatisticCollection::refreshAggregates()
+{
+    CRuntimeStatisticCollection totals(aggregateKindsMapping);
+    Owned<IBitSet> totalUpdated = createBitSet(aggregateKindsMapping.numStatistics());
+    return statsCollection->refreshAggregates(totals, *totalUpdated);
+}
+
+// Load aggregates in collection from global stats
+void GlobalStatisticCollection::updateAggregates(IWorkUnit *wu)
+{
+    class StatisticsAggregatesWriter : implements IStatisticVisitor
+    {
+        const StatisticsMapping & aggregateKindsMapping;
+        const unsigned numStats;
+        Linked<IWorkUnit> wu;
+    public:
+        StatisticsAggregatesWriter(IWorkUnit * _wu, const StatisticsMapping & _aggregateKindsMapping): wu(_wu), aggregateKindsMapping(_aggregateKindsMapping), numStats(aggregateKindsMapping.numStatistics()) {}
+
+        virtual bool visitScope(const IStatisticCollection & cur)
         {
-            it->next();
+            switch (cur.queryScopeType())
+            {
+            case SSTglobal:
+            case SSTworkflow:
+            case SSTgraph:
+                for (unsigned i=0; i<numStats; ++i)
+                {
+                    StatisticKind kind = aggregateKindsMapping.getKind(i);
+                    stat_type value;
+                    if (cur.getStatistic(kind, value) && value)
```

Comment:
you do sometimes want to set if 0, it depends, should be using includeStatisticIfZero(kind) I think.